src/allmydata/web/directory.py

```diff
@@ -943,8 +967,128 @@ def render_get_name(self, ctx, data):
         ctx.tag.attributes['value'] = name
         return ctx.tag
 
-class ManifestResults(MultiFormatPage, ReloadMixin):
-    docFactory = getxmlfile("manifest.xhtml")
+
+class ReloadableMonitorElement(Element):
+    """
+    Like 'ReloadMixin', but for twisted.web.template style. This
+    provides renderers for "reload" and "refesh" and a self.monitor
+    attribute (which is an instance of IMonitor)
+    """
+    refresh_time = timedelta(seconds=60)
+
+    def __init__(self, monitor):
+        self.monitor = monitor
+        super(ReloadableMonitorElement, self).__init__()
+
+    @renderer
+    def refresh(self, req, tag):
+        if self.monitor.is_finished():
+            return u""
+        tag.attributes[u"http-equiv"] = u"refresh"
+        tag.attributes[u"content"] = u"{}".format(self.refresh_time.seconds)
+        return tag
+
+    @renderer
+    def reload(self, req, tag):
+        if self.monitor.is_finished():
+            return u""
+        reload_url = URL.from_text(u"{}".format(req.path))
+        cancel_button = tags.form(
+            [
+                tags.input(type=u"submit", value=u"Cancel"),
+            ],
+            action=reload_url.replace(query={u"t": u"cancel"}).to_uri().to_text(),
+            method=u"POST",
+            enctype=u"multipart/form-data",
+        )
+
+        return tag([
+            u"Operation still running: ",
+            tags.a(
+                u"Reload",
+                href=reload_url.replace(query={u"output": u"html"}).to_uri().to_text(),
+            ),
+            cancel_button,
+        ])
+
+
+def _slashify_path(path):
+    """
+    Converts a tuple from a 'manifest' path into a string with slashes
+    in it
+    """
+    if not path:
+        return ""
+    return "/".join([p.encode("utf-8") for p in path])
+
+
+def _cap_to_link(root, path, cap):
+    """
+    Turns a capability-string into a WebAPI link tag
+
+    :param text root: the root piece of the URI
+
+    :param text cap: the capability-string
+
+    :returns: something suitable for `IRenderable`, specifically
+        either a valid local link (tags.a instance) to the capability
+        or an empty string.
+    """
+    # TODO: we need a clean consistent way to get the type of a cap string
```

Comment:
maybe `allmydata.uri.from_string(...)`?  Then there's a bunch of `is_xyz` methods and some interfaces that can be checked.

Reply:
I *think* what I meant by that comment was wanting e.g. `is_chk(u"blarg")` or so .. but yeah, I could use `.from_string` and make it an `isinstance()` instead of a `startswith`, which is probably better. Thanks :)